File: java/tour-of-restate/src/main/java/dev/restate/tour/auxiliary/PaymentClient.java
/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/tour-of-restate
 */

package dev.restate.tour.auxiliary;

public class PaymentClient {
    private static int i = 0;

    public static PaymentClient get() {
        return new PaymentClient();
    }

    public boolean call(String idempotencyKey, double amount) {
        System.out.println("Payment call succeeded for idempotency key " + idempotencyKey + " and amount " + amount);
        // do the call
        return true;
    }

    public boolean failingCall(String idempotencyKey, double amount) {
        if (i >= 2) {
            System.out.println("Payment call succeeded for idempotency key " + idempotencyKey + " and amount " + amount);
            i = 0;
            return true;
        } else {
            System.out.println("Payment call failed for idempotency key " + idempotencyKey + " and amount " + amount + ". Retrying...");
            i = i + 1;
            throw new IllegalStateException("Payment call failed");
        }
    }
}

File: typescript/tutorials/tour-of-restate-typescript/src/auxiliary/payment_client.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

let i = 0;
export class PaymentClient {
  public static get() {
    return new PaymentClient();
  }
  async call(idempotencyKey: string, amount: number): Promise<boolean> {
    console.log(
      `Payment call succeeded for idempotency key ${idempotencyKey} and amount ${amount}`,
    );
    // do the call
    return true;
  }

  async failingCall(idempotencyKey: string, amount: number): Promise<boolean> {
    if (i >= 2) {
      console.log(
        `Payment call succeeded for idempotency key ${idempotencyKey} and amount ${amount}`,
      );
      i = 0;
      return true;
    } else {
      console.log(
        `Payment call failed for idempotency key ${idempotencyKey} and amount ${amount}. Retrying...`,
      );
      i = i + 1;
      throw new Error("Payment call failed");
    }
  }
}
